### PR TITLE
su - USER command execution support & change transport_options to attr_accessor

### PR DESCRIPTION
**Pull request by vsingh-msys**
## Description

 - Modify transport options whenever required.
 - su - USER command execution using -c require a password ie. `su - USER -c 'ls;date;'`
 - Append the `su_password` option if command execution required a password.

## Related Issue

Fixes: https://github.com/inspec/train/issues/633

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.

Signed-off-by: Vivek Singh <vivek.singh@msystechnologies.com>

### CODE REVIEW

File: lib/train/transports/ssh_connection.rb [hunk 1]
```diff
@@ -316,8 +316,12 @@ def execute_on_channel(cmd)
         channel.exec(cmd) do |_, success|
           abort "Couldn't execute command on SSH." unless success
           channel.on_data do |_, data|
-            yield(data) if block_given?
-            stdout += data
+            if data == "Password: "
+              channel.send_data "#{@transport_options[:su_password]}\n"
```
Comment (lamont-granquist) [inline lib/train/transports/ssh_connection.rb:320]: This should use the accessor.  Also it looks like any time the data contains "Password: " the su_password will be output, even if it is not set?  It also isn't clear to me why this isn't just done
in the caller by passing a block that looks for "Password: " and outputs the su_password, and why this needs to go into train.

File: lib/train/transports/ssh_connection.rb [hunk 2]
```diff
@@ -316,12 +316,12 @@ def execute_on_channel(cmd)
         channel.exec(cmd) do |_, success|
           abort "Couldn't execute command on SSH." unless success
           channel.on_data do |_, data|
-            yield(data) if block_given?
+            yield(data, channel) if block_given?
```
Comment (vsingh-msys) [inline lib/train/transports/ssh_connection.rb:319]: Now we can consume this on the train side with password the data block
```
require 'train'
train = Train.create('ssh', host: 'node1.test', port: 22, user: 'vagrant', key_files: './.vagrant/machines/default/virtualbox/private_key', pty: true)
conn = train.connection

password = "PASSWORD"
res = conn.run_command("su - root -c 'date;'") do |data, ch|
	ch.send_data("#{password}\n") if data.match?("Password:")
end

puts "STDOUT: #{res.stdout}"
puts "STDERR: #{res.stderr}"
```

File: lib/train/transports/ssh_connection.rb [hunk 3]
```diff
@@ -316,12 +316,12 @@ def execute_on_channel(cmd)
...
             stdout += data
           end
 
           channel.on_extended_data do |_, _type, data|
-            yield(data) if block_given?
+            yield(data, channel) if block_given?
```
Comment (vsingh-msys) [inline lib/train/transports/ssh_connection.rb:324]: Added to maintain the parity with `on_data` block.

File: lib/train/transports/ssh_connection.rb [hunk 4]
```diff
@@ -242,6 +242,16 @@ def run_command_via_connection(cmd, &data_handler)
...
 
+      # An interactive console might contain the STDERR in STDOUT
+      # concat both outputs for non-zero exit status. 
+      output = "#{stdout} #{stderr}".strip if exit_status != 0
+
+      # Abstract the su - USER authentication failure
+      # raise the Train::UserError and passes message & reason
+      if output && output.match?("su: Authentication failure")
+        raise Train::UserError.new(output, :bad_su_user_password)
+      end
```
Comment (vsingh-msys) [inline lib/train/transports/ssh_connection.rb:253]: Wrap the `su: Authentication failure` run_command error message and raise custom `Train::UserError` with reason `bad_su_user_password`

**Comment by vsingh-msys** [comment]
@clintoncwolfe could you please have a look? Thanks!

**Comment by tas50** [comment]
@clintoncwolfe Any chance you can chime in on this change?

File: lib/train/transports/ssh_connection.rb (hunk 2 above)
Comment (clintoncwolfe) [reply lib/train/transports/ssh_connection.rb:329]: I've tested using the given code and it does work. I also verified that the new block parameter is optional, so any existing code that calls run_command without the `channel` param will not break.

This does put all of the responsibility for handling the password on the calling code.

File: lib/train/transports/ssh_connection.rb [hunk 5]
```diff
@@ -29,8 +29,8 @@ class Train::Transports::SSH
   #
   # @author Fletcher Nichol <fnichol@nichol.ca>
   class Connection < BaseConnection # rubocop:disable Metrics/ClassLength
-    attr_reader :hostname
-    attr_reader :transport_options
+    attr_reader   :hostname
+    attr_accessor :transport_options
```
Comment (clintoncwolfe) [inline lib/train/transports/ssh_connection.rb:33]: Could you help me understand why this change is needed? I don't see anywhere in the PR what the transport_option ivar is directly changed, and if you need to change one of its elements you can do so without making it read-write.

Comment (lamont-granquist) [reply lib/train/transports/ssh_connection.rb:33]: Here's the context where the goal is to change the pty option to true later in the process:

https://github.com/chef/chef/pull/10410/files#r492238172

I had the same question, but didn't know enough about train to suggest anything better.

Comment (james-stocks) [reply lib/train/transports/ssh_connection.rb:33]: Does this work OK with the ssh code [here](https://github.com/inspec/train/blob/master/lib/train/transports/ssh.rb#L81) that looks at the options to determine if the connection can be re-used or needs restarted?

Comment (vsingh-msys) [reply lib/train/transports/ssh_connection.rb:33]: @james-stocks As we memoinzed the connection object at chef core side https://github.com/chef/chef/blob/0a29892953b07ba1fded119f5aef364e14c7dd07/lib/chef/knife/bootstrap/train_connector.rb#L51-L59, won't be a problem here, but not sure about other impacts.
